lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,188 @@
+
+visSharedMemWriter::visSharedMemWriter(Config& config, const std::string& unique_name, bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&visSharedMemWriter::main_thread, this)) {
+
+        // Fetch any simple configuration
+        root_path = config.get_default<std::string>(unique_name, "root_path", "/dev/shm/");
+        sem_name = config.get_default<std::string>(unique_name, "sem_name", "kotekan");
+        fname_access_record = config.get_default<std::string>(unique_name, "fname_access_record", "calBufferAccessRecord");
+        fname_buf = config.get_default<std::string>(unique_name, "fname_buf", "calBuffer");
+        ntime = config.get_default<size_t>(unique_name, "nsamples", 512);
+
+        // Setup the input vector
+        in_buf = get_buffer("in_buf");
+        register_consumer(in_buf, unique_name.c_str());
+
+        // Check if any of the old buffer files exist
+        DEBUG("Checking for and removing old buffer files...");
+        check_remove(root_path + "sem." + sem_name);
+        check_remove(root_path + fname_access_record);
+        check_remove(root_path + fname_buf);
+
+        // Create the semaphore, and gain first access to it
+        sem = sem_open(
+                sem_name.c_str(),
+                (O_CREAT | O_EXCL),
+                (S_IRUSR | S_IWUSR),
+                1
+            );
+
+        if (sem == SEM_FAILED) {
+            perror("sem_open");
+            exit(errno);
+        }
+
+        INFO("Semaphore created.\n");
+
+        // Acquire semaphore until shared memory is created
+        CHECK(sem_wait(sem));
+
+
+}
+
+// make a deconstructor in which to deconstruct semaphores and shared memory
+visSharedMemWriter::~visSharedMemWriter() {
+    // make sure to unlink the semaphore and unmap the mappings
+}
+
+// takes the name of a shared memory address opens it, and maps the memory to the provided address pointer
+template <typename T>
+T visSharedMemWriter::assign_memory(std::string shm_name, int shm_size, T addr) {
+        int fd = shm_open(shm_name.c_str(), (O_CREAT | O_RDWR), (S_IRUSR | S_IWUSR));
+
+        if (fd == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Cannot open shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+
+        // Resize object to hold buffer
+        CHECK(ftruncate(fd, shm_size));
+        INFO("Resized to %ld bytes\n", (long) shm_size);
+
+        addr = (T) mmap(nullptr, shm_size, (PROT_READ | PROT_WRITE), MAP_SHARED, fd, 0);
+        if (addr == MAP_FAILED) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to map shm {:s} to memory: {:s}."), shm_name, strerror(errno)));
+        }
+
+        // fd is no longer needed
+        CHECK(close(fd));
+
+        return addr;
+}
+
+void visSharedMemWriter::write_to_memory(const visFrameView& frame, size_t index) {
+
+        CHECK(sem_wait(sem));
```

Comment:
Some comments to explain what this is all about would be nice (only if we actually plan to merge this code).